app/models/user.rb

```diff
@@ -0,0 +1,39 @@
+class User < ActiveRecord::Base
+  has_many :article, dependent: :destroy
+  has_many :comment
+  belongs_to :role
+
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable and :omniauthable
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :trackable, :validatable, :omniauthable, :omniauth_providers => [:facebook]
+
+  before_create :set_default_role
+
+  def is_admin?
+    persisted? and (!role.nil?) and (role.name == "admin") 
+  end
+
+  def self.from_omniauth(auth)
+    where(provider: auth.provider, uid: auth.uid).first_or_create do |user|
```

Comment:
there's a couple of cases that you missed here
1. what if the user signed up with their email earlier and then did facebook afterwards? 
something like this: 
```
where(provider: auth.provider, uid: auth.uid).last || where('lower(email)' => auth.info.email.downcase).last
```

or even better; a single where query
```
where('(provider = ? and uid = ?) or lower(email) = ?' , auth.provider, auth.uid, auth.info.email.downcase).last
```
I also handled case sensitivity on your email address there.